lib/internal/streams/pipeline.js

```diff
@@ -54,6 +56,23 @@ function destroyer(stream, reading, writing, final, callback) {
       stream
         .once('end', _destroy)
         .once('error', _destroy);
+    } else if (
+      !err &&
+      (!final || !stream.readable) &&
+      state &&
+      state.autoDestroy &&
+      state.emitClose &&
+      state.closed === false
+    ) {
+      // TODO (ronag): Improve soft detection to include core modules and
+      // common ecosystem modules that do properly emit 'close' but fail
+      // the generic check above.
+
+      // 'eos' will invoke the callback early on 'finish' and 'end' due to
+      // compat reasons. Try to soft detect whether 'close' will be emitted.
```

Comment:
What are those compatibility reasons? Or in other words, what would break if we were to fix this in eos instead?

Reply:
> what would break if we were to fix this in eos instead?

Not sure, it does change timing, so it might be more breaking than this but I can't think of anything specific. Good point. @mcollina what do you think?

Reply:
@vweevers: it would break e.g. `pipeline` pre #32051, so there is a risk of breaking other stuff.

Reply:
That's unfortunate. Ideally `stream.finished` behaves the same as `stream.pipeline`.

Reply:
@vweevers Updated according to your suggestion. We can try a CITGM run once #32051 has landed.